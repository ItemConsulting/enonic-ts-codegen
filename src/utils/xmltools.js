"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
var change_case_1 = require("change-case");
var path = require("path");
var xmldom = require("xmldom");
var xpathutils_1 = require("./xpathutils");
var xpath = require("xpath");
exports.MissingFieldNameError = "A field is missing a name attribute";
/**
 * createInterface parses an xml string and generates code for a TypeScript interface.
 */
function createInterface(interfaceName, xml) {
    return formatInterface(parseXML(interfaceName, xml));
}
exports.createInterface = createInterface;
/**
 * generateInterfaceName generates a name for a TypeScript interface from a filename.
 * @param filename
 */
function generateInterfaceName(filename) {
    return change_case_1.pascalCase(path.basename(filename, path.extname(filename)));
}
exports.generateInterfaceName = generateInterfaceName;
function formatInterface(iface) {
    var fields = iface.fields.map(formatField).join("\n\n");
    return "export interface " + iface.name + " {\n" + fields + "\n};\n";
}
function formatField(field) {
    var optional = field.optional ? "?" : "";
    var comment = field.comment ? formatComment(field.comment) : "";
    var subfields = field.subfields && field.subfields.length > 0
        ? "<{\n" +
            field.subfields
                .map(formatField)
                .join("\n\n")
                .split("\n")
                .map(function (line) { return (line.length > 0 ? "  " + line : line); })
                .join("\n") +
            "\n  }>"
        : "";
    return comment + "  " + field.name + optional + ": " + field.type + subfields;
}
function formatComment(comment) {
    var commentLines = comment
        .split("\n")
        .map(function (line) { return "   * " + line; })
        .join("\n");
    return "  /**\n" + commentLines + "\n   */\n";
}
function parseXML(name, xml) {
    var doc = new xmldom.DOMParser().parseFromString(xml);
    var form = xpathutils_1.evaluate("//form", doc).iterateNext();
    return { name: name, fields: parseForm(form) };
}
exports.parseXML = parseXML;
function parseForm(form) {
    return form
        ? __spreadArrays(getInputFields(form), getFieldSetItems(form), getItemSetFields(form)) : [];
}
function getInputFields(node) {
    var inputs = xpathutils_1.evaluate("./input", node);
    return xpathutils_1.mapXpathResult(inputs, createFieldFromInput);
}
// FieldSet is a flat structure with no nesting.
function getFieldSetItems(node) {
    var fieldSets = xpathutils_1.evaluate("./field-set", node);
    return xpathutils_1.flatmapXpathResult(fieldSets, function (node) {
        return xpathutils_1.mapXpathResult(xpathutils_1.evaluate("./items/input", node), createFieldFromInput);
    });
}
// ItemSet is a nested structure.
function getItemSetFields(node) {
    var itemSets = xpathutils_1.evaluate("./item-set", node);
    return xpathutils_1.mapXpathResult(itemSets, function (node) {
        var nameAttr = xpath.select1("@name", node);
        var minimumOccurrencesAttr = xpath.select1("./occurrences/@minimum", node);
        var name = nameAttr.value;
        var type = "Array";
        var optional = minimumOccurrencesAttr
            ? minimumOccurrencesAttr.value === "0"
            : true;
        var items = xpathutils_1.evaluate("./items", node).iterateNext();
        var subfields = __spreadArrays(getInputFields(items), getFieldSetItems(items), getItemSetFields(items));
        return { name: name, type: type, optional: optional, subfields: subfields };
    });
}
function createFieldFromInput(input) {
    var nameAttr = xpath.select1("@name", input);
    if (!nameAttr) {
        throw exports.MissingFieldNameError;
    }
    var minimumOccurrencesAttr = xpath.select1("./occurrences/@minimum", input);
    var name = nameAttr.value;
    var comment = xpath.select1("string(./label)", input);
    var optional = minimumOccurrencesAttr
        ? minimumOccurrencesAttr.value === "0"
        : true;
    var type = "string";
    return { name: name, type: type, comment: comment, optional: optional };
}
