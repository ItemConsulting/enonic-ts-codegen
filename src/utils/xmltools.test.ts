import * as xmltools from "./xmltools";

const xml = {
  employee: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input type="TextLine" name="firstName">
      <label>First name</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
    <input type="TextLine" name="lastName">
      <label>Last name</label>
      <occurrences minimum="0" maximum="1"/>
    </input>
    <item-set name="contactInfo">
      <label>Contact Info</label>
      <occurrences minimum="0" maximum="0"/>
      <items>
        <input name="label" type="TextLine">
          <label>Label</label>
          <occurrences minimum="0" maximum="1"/>
        </input>
        <input name="phoneNumber" type="TextLine">
          <label>Phone Number</label>
          <occurrences minimum="0" maximum="1"/>
        </input>
      </items>
    </item-set>
  </form>
  </content-type>
  `,

  simple: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input type="TextLine" name="firstName">
      <label>First name</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
    <input type="TextLine" name="lastName">
      <label>Last name</label>
      <occurrences minimum="0" maximum="1"/>
    </input>
  </form>
  </content-type>\n`,

  simpleMissingName: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input type="TextLine">
      <label>First name</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
  </form>
  </content-type>\n`,

  simpleCaseIncentiveType: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input name="gdprSigned" type="cHeckbOx">
      <label>GDPR Signed</label>
    </input>
  </form>
  </content-type>\n`,

  simpleMissingComment: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input type="TextLine" name="firstName">
      <occurrences minimum="1" maximum="1"/>
    </input>
  </form>
  </content-type>\n`,

  simpleMissingOccurrences: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input type="TextLine" name="firstName">
    </input>
    <input type="TextLine" name="lastName">
      <occurrences maximum="1"/>
    </input>
  </form>
  </form>
  </content-type>\n`,

  itemSet: `<content-type>
<display-name>Employee</display-name>
<super-type>base:structured</super-type>
<form>
  <item-set name="contactInfo">
    <label>Contact Info</label>
    <occurrences minimum="0" maximum="0"/>
    <items>
      <input name="label" type="TextLine">
        <label>Label</label>
        <occurrences minimum="0" maximum="1"/>
      </input>
      <input name="phoneNumber" type="TextLine">
        <label>Phone Number</label>
        <occurrences minimum="0" maximum="1"/>
      </input>
      <item-set name="contactInfo">
        <label>Contact Info</label>
        <occurrences minimum="0" maximum="0"/>
        <items>
          <input name="phoneNumber" type="TextLine">
            <label>Phone Number</label>
            <occurrences minimum="0" maximum="1"/>
          </input>
        </items>
      </item-set>
    </items>
  </item-set>
</form>
</content-type>\n`,

  fieldSet: `<content-type>
<display-name>Employee</display-name>
<super-type>base:structured</super-type>
<form>
  <field-set name="Juryvotes">
    <label>Juryavgj√∏relser</label>
    <items>
        <input type="TextArea" name="juryvote1">
            <label>Runde 1</label>
            <occurrences minimum="0" maximum="1"/>
        </input>
        <input type="TextArea" name="juryvote2">
            <label>Runde 2</label>
            <occurrences minimum="0" maximum="1"/>
        </input>
    </items>
  </field-set>
</form>
</content-type>\n`,

  minimalSite: `<site>
  <form/>
</site>`,

  simpleSite: `<site>
  <processors> 
    <response-processor name="tracker" order="10"/>  
  </processors>
  <mappings> 
     <mapping controller="/site/foobar/api.js" order="10">
       <pattern>/api/v\d+/.*</pattern>
     </mapping>
     <mapping controller="/site/pages/default/default.js">
       <match>type:'portal:fragment'</match>
     </mapping>
   </mappings>
   <form> 
    <input type="TextLine" name="company">
      <label>Company</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
  </form>
  <x-data name="seo-settings"/> 
</site>`,

  simplePage: `<page>
  <display-name i18n="component.page.name">My first page</display-name>
  <description>Front page of our site</description>
  <form>
    <input type="TextLine" name="pageName">
      <label>Page Name</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
  </form>
  <regions>
    <region name="main"/>
  </regions>
</page>`,

  simplePart: `<part>
  <display-name i18n="employeeForm.displayName">Employee Form</display-name>
  <description i18n="employeeForm.description">Edit the information of an employee</description>
  <form>
    <input type="TextLine" name="partPart">
      <label>Part of a part</label>
      <occurrences minimum="0" maximum="10"/>
    </input>
  </form>
</part>`,

  missingForm: `<part>
  <display-name i18n="groupForm.displayName">Group form</display-name>
  <description i18n="groupForm.description">Edit description about tahe group</description>
</part>`,

  mixinUser: `<content-type>
  <display-name>Using mixins</display-name>
  <form>
    <input type="TextLine" name="firstName">
      <label>First Name</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
    <input type="TextLine" name="lastName">
      <label>Last Name</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
    <mixin name="address"/>
  </form>
</content-type>`,

  mixin: `<mixin>
  <display-name>Full address</display-name>
  <form>
    <input type="TextLine" name="addressLine">
      <label>Street address</label>
      <occurrences minimum="0" maximum="2"/>
    </input>
    <input type="TextLine" name="city">
      <label>City</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
    <input type="TextLine" name="zipCode">
      <label>Zip code</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
    <input type="TextLine" name="country">
      <label>Country</label>
      <occurrences minimum="1" maximum="1"/>
    </input>
  </form>
</mixin>`,

  noForm: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
</content-type>\n`,

  emptyForm: `<content-type>
  <display-name>Employee</display-name>
  <super-type>base:structured</super-type>
  <form></form>
</content-type>\n`,

  contentSelectorSingle: `<content-type>
  <display-name>Group</display-name>
  <super-type>base:structured</super-type>
  <form>
  <input name="fabTarget" type="ContentSelector">
    <label>FAB button target</label>
    <occurrences minimum="0" maximum="1"/>
  </input>
  </form>
</content-type>\n`,

  contentSelectorMultiple: `<content-type>
  <display-name>Group</display-name>
  <super-type>base:structured</super-type>
  <form>
    <input name="members" type="ContentSelector">
      <label>Members</label>
      <occurrences minimum="0" maximum="2"/>
      <config>
        <allowContentType>employee</allowContentType>
      </config>
    </input>
  </form>
</content-type>\n`,

  checkbox: `<content-type>
  <display-name>Using mixins</display-name>
  <form>
    <input name="gdprSigned" type="CheckBox">
      <label>GDPR Signed</label>
    </input>
  </form>
</content-type>\n`,


  combobox: `<content-type>
  <display-name>Using mixins</display-name>
  <form>
    <input name="invite" type="ComboBox">
    <label>Invited</label>
    <occurrences minimum="0" maximum="1"/>
      <config>
        <option value="Yes">Yes</option>
        <option value="No">No</option>
        <option value="Maybe">Maybe</option>
      </config>
    </input>
  </form>
</content-type>\n`,

  optionSet: `<content-type>
  <display-name>Using mixins</display-name>
  <form>
    <option-set name="checkOptionSet">
      <label i18n="checkOptionSet.label">Multi-selection OptionSet</label>
      <expanded>true</expanded> 
      <occurrences minimum="1" maximum="1"/> 
      <help-text>You can select up to 2 options</help-text>
      <options minimum="1" maximum="2">    
        <option name="option_1">  
          <label i18n="checkOptionSet.option_1.label">Option 1</label>  
          <help-text i18n="checkOptionSet.option_1.help-text">Help text for Option 1</help-text>  
        </option>
        <option name="option_2">
          <label i18n="checkOptionSet.option_2.label">Option 2</label>
          <default>true</default> 
          <items> 
            <input name="contentSelector" type="ContentSelector">
              <label>Content selector</label>
              <occurrences minimum="0" maximum="0"/>
              <config/>
            </input>
          </items>
        </option>
        <option name="option_3">
          <label>Option 3</label>
          <help-text>Help text for Option 3</help-text>
          <items>
            <input name="textarea" type="TextArea">
              <label>Text Area</label>
              <occurrences minimum="0" maximum="1"/>
            </input>
            <input name="long" type="Long">
              <label>Long</label>
              <indexed>true</indexed>
              <occurrences minimum="0" maximum="1"/>
            </input>
          </items>
        </option>
      </options>
    </option-set>
  </form>
  </content-type>\n`,

  radioButton: `
    <content-type>
      <display-name>Using mixins</display-name>
      <form>
        <input name="myradiobutton" type="RadioButton">
          <label>My RadioButton</label>
          <occurrences minimum="0" maximum="1"/> 
          <config>
            <option value="one">Option One</option> 
            <option value="two">Option Two</option>
          </config>
          <default>one</default> 
        </input>
      </form>
  </content-type>\n`
};

describe("parseXML", () => {
  test("sets the name field", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.simple);
    expect(tsInterface.name).toBe("Employee");
  });

  test("parses all inputs", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.simple);
    expect(tsInterface.fields.length).toBe(2);
  });

  test("parses field names", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.simple);
    expect(tsInterface.fields[0].name).toBe("firstName");
    expect(tsInterface.fields[1].name).toBe("lastName");
  });

  test("parses no form as 0 fields", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.missingForm);
    expect(tsInterface.fields).toHaveLength(0);
  });

  test("parses optional inputs", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.simple);
    expect(tsInterface.fields[0].optional).toBe(false);
    expect(tsInterface.fields[1].optional).toBe(true);
  });

  test("parses inputs with missing name", () => {
    expect(() => {
      xmltools.parseXml("Employee", xml.simpleMissingName);
    }).toThrow();
  });

  test("parses inputs with missing comment", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.simpleMissingComment);
    expect(tsInterface.fields[0].comment).toBe("");
  });

  test("parses optional inputs with missing occurrences", () => {
    // Occurences defaults to minimum=0, maximum=1
    // See: https://developer.enonic.com/docs/xp/stable/cms/schemas#input_types
    const tsInterface = xmltools.parseXml(
      "Employee",
      xml.simpleMissingOccurrences
    );
    expect(tsInterface.fields[0].optional).toBe(true);
    expect(tsInterface.fields[1].optional).toBe(true);
  });

  test("parses FieldSets as a flat structure", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.fieldSet);
    expect(tsInterface.fields.length).toBe(2);
  });

  test("parses ItemSets as a nested structure", () => {
    const tsInterface = xmltools.parseXml("Employee", xml.itemSet);
    expect(tsInterface.fields.length).toBe(1);
    expect(tsInterface.fields[0].subfields.length).toBe(3);
    expect(tsInterface.fields[0].subfields[2].subfields.length).toBe(1);
  });

  test("parses ContentSelector as string", () => {
    const tsInterface = xmltools.parseXml(
      "ContentSelectorExample",
      xml.contentSelectorSingle
    );
    const field = tsInterface.fields[0];
    expect(field.type).toBe("string");
  });

  test("parses ContentSelector as Array<string>", () => {
    const tsInterface = xmltools.parseXml(
      "ContentSelectorExample",
      xml.contentSelectorMultiple
    );
    const field = tsInterface.fields[0];
    expect(field.type).toBe("Array<string>");
  });

  test("parses CheckBox as boolean", () => {
    const tsInterface = xmltools.parseXml("CheckBoxExample", xml.checkbox);
    const field = tsInterface.fields[0];
    expect(field.type).toBe("boolean");
  });

  test("parses Input types regardless of case", () => {
    const tsInterface = xmltools.parseXml("CaseIncentiveTypeExample", xml.simpleCaseIncentiveType);
    const field = tsInterface.fields[0];
    expect(field.type).toBe("boolean");
  });

  test("parses the minimal site", () => {
    const tsInterface = xmltools.parseXml("MySite", xml.minimalSite);
    expect(tsInterface.name).toBe("MySite");
    expect(tsInterface.fields).toHaveLength(0);
  });

  test("parses a simple site", () => {
    const tsInterface = xmltools.parseXml("MySite", xml.simpleSite);
    expect(tsInterface.name).toBe("MySite");
    expect(tsInterface.fields).toHaveLength(1);

    const field = tsInterface.fields[0];
    expect(field.name).toBe("company");
    expect(field.comment).toBe("Company");
    expect(field.optional).toBe(false);
  });

  test("parses a simple page", () => {
    const tsInterface = xmltools.parseXml("MyPage", xml.simplePage);
    expect(tsInterface.name).toBe("MyPage");
    expect(tsInterface.fields).toHaveLength(1);

    const field = tsInterface.fields[0];
    expect(field.name).toBe("pageName");
    expect(field.optional).toBe(false);
    expect(field.comment).toBe("Page Name");
  });

  test("parses a simple part", () => {
    const tsInterface = xmltools.parseXml(
      "SevenElevenWasAPartTimeJob",
      xml.simplePart
    );
    expect(tsInterface.name).toBe("SevenElevenWasAPartTimeJob");
    expect(tsInterface.fields).toHaveLength(1);

    const field = tsInterface.fields[0];
    expect(field.name).toBe("partPart");
    expect(field.optional).toBe(true);
    expect(field.comment).toBe("Part of a part");
  });

  test("parses mixins", () => {
    const tsInterface = xmltools.parseXml("MixinExample", xml.mixinUser);
    const mixins = tsInterface.fields.filter(field => field.type === "mixin");
    expect(mixins).toHaveLength(1);
    expect(mixins[0].name).toBe("address");
    expect(mixins[0].type).toBe("mixin");
  });
});

describe("InterfaceGenerator", () => {
  test("NewInterfaceGenerator creates a new InterfaceGenerator", () => {
    const generator = xmltools.NewInterfaceGenerator();
    expect(generator).toBeDefined();
    expect(generator).not.toBeNull();
  });

  test("throws when no mixin exists", () => {
    const generator = xmltools.NewInterfaceGenerator();
    expect(() => {
      generator.createInterface("MixedIn", xml.mixinUser);
    }).toThrow();
  });

  test("generates the correct input code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface("Simple", xml.simple);
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct FieldSet code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface("FieldSetUser", xml.fieldSet);
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct ItemSet code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface("ItemSetUser", xml.itemSet);
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct ContentSelector code", () => {
    const generator = xmltools.NewInterfaceGenerator();

    const tsInterfaceSingle = generator.createInterface(
      "ContentSelectorExample",
      xml.contentSelectorSingle
    );
    expect(tsInterfaceSingle).toMatchSnapshot();

    const tsInterfaceMultiple = generator.createInterface(
      "ContentSelectorExample",
      xml.contentSelectorMultiple
    );
    expect(tsInterfaceMultiple).toMatchSnapshot();
  });

  test("generates the correct OptionSet code",() => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface(
      "optionSet",
      xml.optionSet
    );
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct CheckBox code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface(
      "CheckBoxExample",
      xml.checkbox
    );
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct ComboBox code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface(
      "ComboBoxExample",
      xml.combobox
    );
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct RadioButton code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface(
      "RadioButtonExample",
      xml.radioButton
    );
    expect(tsInterface).toMatchSnapshot();
  });

  test("generates the correct mixin code", () => {
    const generator = xmltools.NewInterfaceGenerator();
    generator.addMixin("address", xml.mixin);
    const tsInterface = generator.createInterface("MixedIn", xml.mixinUser);
    expect(tsInterface).toMatchSnapshot();
  });

  test("returns empty string when no form can be found", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface("ShouldBeNull", xml.noForm);
    expect(tsInterface).toBe("");
  });

  test("returns empty string when form is empty", () => {
    const generator = xmltools.NewInterfaceGenerator();
    const tsInterface = generator.createInterface(
      "ShouldBeNull",
      xml.emptyForm
    );
    expect(tsInterface).toBe("");
  });
});

describe("generateInterfaceName", () => {
  test("sets the interface name in PascalCase", () => {
    const filenames = [
      "my_little_interface",
      "myLittleInterface",
      "my-little-interface"
    ];
    for (const filename of filenames) {
      const iname = xmltools.generateInterfaceName(filename);
      expect(iname).toBe("MyLittleInterface");
    }
  });

  test("drops the file extension", () => {
    const filename = "my-little-interface.xml";
    const iname = xmltools.generateInterfaceName(filename);
    expect(iname).toBe("MyLittleInterface");
  });

  test("drops the file path", () => {
    const filename = "/path/to/my-little-interface.xml";
    const iname = xmltools.generateInterfaceName(filename);
    expect(iname).toBe("MyLittleInterface");
  });
});
