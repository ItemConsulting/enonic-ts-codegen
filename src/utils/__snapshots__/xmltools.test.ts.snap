// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterfaceGenerator generates the correct CheckBox code 1`] = `
"export interface CheckBoxExample {
  /**
   * GDPR Signed
   */
  gdprSigned: boolean;
}
"
`;

exports[`InterfaceGenerator generates the correct ComboBox code 1`] = `
"export interface ComboBoxExample {
  /**
   * Invited
   */
  invite?: \\"Yes\\" | \\"No\\" | \\"Maybe\\";
}
"
`;

exports[`InterfaceGenerator generates the correct ContentSelector code 1`] = `
"export interface ContentSelectorExample {
  /**
   * FAB button target
   */
  fabTarget?: string;
}
"
`;

exports[`InterfaceGenerator generates the correct ContentSelector code 2`] = `
"export interface ContentSelectorExample {
  /**
   * Members
   */
  members?: Array<string>;
}
"
`;

exports[`InterfaceGenerator generates the correct FieldSet code 1`] = `
"export interface FieldSetUser {
  /**
   * Runde 1
   */
  juryvote1?: string;

  /**
   * Runde 2
   */
  juryvote2?: string;
}
"
`;

exports[`InterfaceGenerator generates the correct ItemSet code 1`] = `
"export interface ItemSetUser {
  /**
   * Contact Info
   */
  contactInfo?: Array<{
    /**
     * Label
     */
    label?: string;

    /**
     * Phone Number
     */
    phoneNumber?: string;

    /**
     * Contact Info
     */
    contactInfo?: Array<{
      /**
       * Phone Number
       */
      phoneNumber?: string;
    }>;
  }>;
}
"
`;

exports[`InterfaceGenerator generates the correct OptionSet code 1`] = `
"export interface optionSet {
  /**
   * Multi-selection OptionSet
   */
  checkOptionSet: {
    /**
     * Selected
     */
    _selected: string | Array<string>;

    /**
     * Option 1
     */
    option_1?: {

    };

    /**
     * Option 2
     */
    option_2?: {
      /**
       * Content selector
       */
      contentSelector?: Array<string>;
    };

    /**
     * Option 3
     */
    option_3?: {
      /**
       * Text Area
       */
      textarea?: string;

      /**
       * Long
       */
      long?: string;
    };
  };
}
"
`;

exports[`InterfaceGenerator generates the correct RadioButton code 1`] = `
"export interface RadioButtonExample {
  /**
   * My RadioButton
   */
  myradiobutton?: \\"one\\" | \\"two\\";
}
"
`;

exports[`InterfaceGenerator generates the correct input code 1`] = `
"export interface Simple {
  /**
   * First name
   */
  firstName: string;

  /**
   * Last name
   */
  lastName?: string;
}
"
`;

exports[`InterfaceGenerator generates the correct mixin code 1`] = `
"export interface MixedIn {
  /**
   * First Name
   */
  firstName: string;

  /**
   * Last Name
   */
  lastName: string;

  /**
   * Street address
   */
  addressLine?: string;

  /**
   * City
   */
  city: string;

  /**
   * Zip code
   */
  zipCode: string;

  /**
   * Country
   */
  country: string;
}
"
`;

exports[`InterfaceGenerator generates the correct mixin code for multilevel 1`] = `
"export interface MixedIn {
  /**
   * Street address
   */
  addressLine?: string;

  /**
   * City
   */
  city: string;

  /**
   * Zip code
   */
  zipCode: string;

  /**
   * Country
   */
  country: string;

  /**
   * Contact Info
   */
  contact_info?: Array<{
    /**
     * Label
     */
    label?: string;

    /**
     * Street address
     */
    addressLine?: string;

    /**
     * City
     */
    city: string;

    /**
     * Zip code
     */
    zipCode: string;

    /**
     * Country
     */
    country: string;
  }>;

  /**
   * Multi-selection OptionSet
   */
  checkOptionSet: {
    /**
     * Selected
     */
    _selected: string | Array<string>;

    /**
     * Option 1
     */
    option_1?: {

    };

    /**
     * Option 2
     */
    option_2?: {
      /**
       * Label
       */
      label?: string;

      /**
       * Street address
       */
      addressLine?: string;

      /**
       * City
       */
      city: string;

      /**
       * Zip code
       */
      zipCode: string;

      /**
       * Country
       */
      country: string;
    };

    /**
     * Option 3
     */
    option_3?: {
      /**
       * Street address
       */
      addressLine?: string;

      /**
       * City
       */
      city: string;

      /**
       * Zip code
       */
      zipCode: string;

      /**
       * Country
       */
      country: string;
    };
  };
}
"
`;
